File: frontend/src/types/prediction.ts
/**
 * TypeScript Type Definitions for AI Predictive Incident Detection Engine
 */

export type RiskLevel = "Critical" | "High" | "Medium" | "Low";
export type PredictionStatus =
  | "Active"
  | "Monitoring"
  | "Resolved"
  | "Expired"
  | "False_Positive"
  | "Mitigated"
  | "Dismissed"
  | "Triggered";

export interface MetricConcern {
  name: string;
  current_value: string;
  threshold: string;
  anomaly_trend: string;
  risk_impact: string;
}

export interface MetricForecastPoint {
  horizon: string;
  timestamp: string;
  predicted_value: number;
  lower_bound: number;
  upper_bound: number;
  confidence: number;
}

export interface MetricForecastResponse {
  metric_name: string;
  service: string;
  current_value: number;
  forecast_points: MetricForecastPoint[];
  historical_points: Array<{ timestamp: string; value: number }>;
  model_used: string;
  data_sufficiency: {
    samples: number;
    minimum_required: number;
    sufficient: boolean;
    confidence_factor: number;
  };
  generated_at: string;
}

export interface AnomalyEvent {
  id: string;
  organization_id?: string;
  service: string;
  metric_name: string;
  resource_id?: string;
  value: number;
  baseline_value: number;
  anomaly_score: number;
  severity: "NORMAL" | "WARNING" | "CRITICAL";
  direction: "SPIKE_HIGH" | "DROP_LOW" | "DRIFT" | "NORMAL";
  method: string;
  detected_at: string;
  details?: Record<string, any>;
}

export interface AnomalyDetectionResponse {
  metric_name: string;
  value: number;
  baseline_value: number;
  anomaly_score: number;
  severity: string;
  is_anomaly: boolean;
  direction: string;
  method_used: string;
  z_score: number;
  deviation_percent: number;
  explanation: string;
}

export interface CapacityRiskResponse {
  resource_name: string;
  current_value: number;
  capacity_limit: number;
  exhaustion_threshold: number;
  risk_score: number;
  risk_level: string;
  is_exhaustion_imminent: boolean;
  estimated_time_to_threshold_minutes?: number;
  rate_of_growth_per_minute: number;
  data_status: string;
  summary: string;
  recommended_mitigation: string;
}

export interface Prediction {
  id: string;
  title: string;
  service: string;
  metric_name?: string;
  resource_id?: string;
  environment?: string;
  region: string;
  prediction_score: number;
  failure_probability: number;
  expected_failure_time?: string;
  estimated_time_to_threshold_minutes?: number;
  risk_level: RiskLevel;
  status: PredictionStatus;
  trend_direction?: string;
  trend_strength?: number;
  rate_of_change?: number;
  anomaly_score?: number;
  affected_services: string[];
  likely_root_cause?: string;
  confidence_score: number;
  recommended_preventive_actions: string[];
  triggering_metrics: Record<string, any>;
  data_sufficiency?: Record<string, any>;
  forecast_points?: Array<{
    horizon: string;
    timestamp: string;
    predicted_value: number;
    lower_bound: number;
    upper_bound: number;
    confidence: number;
  }>;
  correlated_signals?: Array<{
    metric: string;
    current_value: number;
    baseline: number;
    anomaly_score: number;
    severity: string;
    direction: string;
    trend: string;
    explanation: string;
  }>;
  analysis_engine?: "gemini" | "local";
  created_at: string;
  updated_at: string;

  // Gemini AI detailed explanations
  ai_explanation?: string;
  ai_metrics_of_concern?: MetricConcern[];
  ai_historical_pattern_comparison?: string;
  ai_possible_impact?: string;
  ai_immediate_preventive_actions?: string[];
  ai_long_term_recommendations?: string[];
}

export interface PredictionListResponse {
  items: Prediction[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface PredictionStats {
  predicted_failures: number;
  high_risk_services: number;
  avg_confidence_percent: number;
  prevented_downtime_hours: number;
}

export interface PredictionAnalytics {
  total_predictions: number;
  active_risks: number;
  critical_risks: number;
  anomaly_events_count: number;
  predicted_failures: number;
  average_confidence: number;
  predictions_by_service: Record<string, number>;
  predictions_by_metric: Record<string, number>;
  predictions_by_risk: Record<string, number>;
}

export interface ServiceRiskItem {
  service: string;
  region: string;
  risk_level: RiskLevel;
  failure_probability: number;
  active_predictions_count: number;
}

export interface InfrastructureRiskHeatmap {
  items: ServiceRiskItem[];
}

export interface PredictionAnalyzePayload {
  services?: string[];
  lookback_hours?: number;
  telemetry_map?: Record<string, number[]>;
}

File: backend/app/schemas/prediction.py
"""
Pydantic v2 Schemas for Predictive AIOps & Anomaly Intelligence Engine.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class RiskLevel(str, Enum):
    CRITICAL = "Critical"
    HIGH = "High"
    MEDIUM = "Medium"
    LOW = "Low"


class PredictionStatus(str, Enum):
    ACTIVE = "Active"
    MONITORING = "Monitoring"
    RESOLVED = "Resolved"
    EXPIRED = "Expired"
    FALSE_POSITIVE = "False_Positive"
    MITIGATED = "Mitigated"
    DISMISSED = "Dismissed"
    TRIGGERED = "Triggered"


class DataSufficiencySchema(BaseModel):
    samples: int
    minimum_required: int
    sufficient: bool
    confidence_factor: float = 1.0


class MetricConcern(BaseModel):
    name: str
    current_value: str
    threshold: str
    anomaly_trend: str
    risk_impact: str


class MetricForecastPointSchema(BaseModel):
    horizon: str  # 5m, 15m, 30m, 1h, 6h, 24h
    timestamp: datetime
    predicted_value: float
    lower_bound: float
    upper_bound: float
    confidence: float


class MetricForecastResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    metric_name: str
    service: str
    current_value: float
    forecast_points: list[MetricForecastPointSchema]
    historical_points: list[dict[str, Any]]
    model_used: str
    data_sufficiency: dict[str, Any]
    generated_at: datetime


class ForecastRequest(BaseModel):
    service: str = Field(default="api-gateway")
    metric_name: str = Field(default="memory_utilization")
    historical_values: list[float] | None = None
    horizons: list[str] = Field(default=["5m", "15m", "30m", "1h", "6h", "24h"])
    step_minutes: int = Field(default=1)


class AnomalyDetectionRequest(BaseModel):
    service: str = Field(default="api-gateway")
    metric_name: str = Field(default="cpu_utilization")
    current_value: float | None = None
    historical_values: list[float] | None = None
    custom_critical_threshold: float | None = None


class AnomalyDetectionResponse(BaseModel):
    metric_name: str
    value: float
    baseline_value: float
    anomaly_score: float
    severity: str
    is_anomaly: bool
    direction: str
    method_used: str
    z_score: float
    deviation_percent: float
    explanation: str


class AnomalyEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID | None = None
    service: str
    metric_name: str
    resource_id: str | None = None
    value: float
    baseline_value: float
    anomaly_score: float
    severity: str
    direction: str
    method: str
    detected_at: datetime
    details: dict[str, Any] | None = None


class BaselineResponse(BaseModel):
    metric_name: str
    window: str
    samples_count: int
    mean: float
    median: float
    standard_deviation: float
    variance: float
    min_value: float
    max_value: float
    p50: float
    p90: float
    p95: float
    p99: float
    rolling_average: float
    rolling_std: float
    data_sufficiency: DataSufficiencySchema
    calculated_at: datetime


class CapacityRiskRequest(BaseModel):
    service: str = Field(default="api-gateway")
    resource_name: str = Field(default="memory_utilization")
    historical_values: list[float] | None = None
    custom_threshold: float | None = None


class CapacityRiskResponse(BaseModel):
    resource_name: str
    current_value: float
    capacity_limit: float
    exhaustion_threshold: float
    risk_score: float
    risk_level: str
    is_exhaustion_imminent: bool
    estimated_time_to_threshold_minutes: float | None = None
    rate_of_growth_per_minute: float
    data_status: str
    summary: str
    recommended_mitigation: str


class PredictionBase(BaseModel):
    title: str = Field(..., min_length=3, max_length=500)
    service: str = Field(default="api-gateway")
    metric_name: str | None = Field(default="memory_utilization")
    resource_id: str | None = None
    environment: str = Field(default="production")
    region: str = Field(default="us-east-1")
    prediction_score: float = Field(default=0.88, ge=0.0, le=1.0)
    failure_probability: float = Field(default=88.5, ge=0.0, le=100.0)
    confidence_score: float = Field(default=0.94, ge=0.0, le=1.0)
    risk_level: RiskLevel = Field(default=RiskLevel.HIGH)
    status: PredictionStatus = Field(default=PredictionStatus.ACTIVE)
    trend_direction: str = Field(default="STABLE")
    trend_strength: float = Field(default=0.0)
    rate_of_change: float = Field(default=0.0)
    anomaly_score: float = Field(default=0.0)
    expected_failure_time: datetime | None = None
    estimated_time_to_threshold_minutes: float | None = None
    affected_services: list[str] = Field(default_factory=list)
    likely_root_cause: str | None = None
    recommended_preventive_actions: list[str] = Field(default_factory=list)
    triggering_metrics: dict[str, Any] = Field(default_factory=dict)
    data_sufficiency: dict[str, Any] | None = Field(default_factory=dict)
    forecast_points: list[dict[str, Any]] | None = Field(default_factory=list)
    correlated_signals: list[dict[str, Any]] | None = Field(default_factory=list)


class PredictionCreate(PredictionBase):
    analysis_engine: str = "local"
    ai_explanation: str | None = None
    ai_metrics_of_concern: list[MetricConcern] | None = Field(default_factory=list)
    ai_historical_pattern_comparison: str | None = None
    ai_possible_impact: str | None = None
    ai_immediate_preventive_actions: list[str] | None = Field(default_factory=list)
    ai_long_term_recommendations: list[str] | None = Field(default_factory=list)


class PredictionUpdate(BaseModel):
    title: str | None = None
    service: str | None = None
    region: str | None = None
    risk_level: RiskLevel | None = None
    status: PredictionStatus | None = None
    recommended_preventive_actions: list[str] | None = None


class PredictionStatusUpdate(BaseModel):
    status: PredictionStatus


class PredictionResponse(PredictionBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
    analysis_engine: str = "local"
    ai_explanation: str | None = None
    ai_metrics_of_concern: list[MetricConcern] | None = Field(default_factory=list)
    ai_historical_pattern_comparison: str | None = None
    ai_possible_impact: str | None = None
    ai_immediate_preventive_actions: list[str] | None = Field(default_factory=list)
    ai_long_term_recommendations: list[str] | None = Field(default_factory=list)


class PredictionListResponse(BaseModel):
    items: list[PredictionResponse]
    total: int
    page: int
    size: int
    pages: int


class PredictionStatsResponse(BaseModel):
    predicted_failures: int
    high_risk_services: int
    avg_confidence_percent: float
    prevented_downtime_hours: float


class PredictionAnalyticsResponse(BaseModel):
    total_predictions: int
    active_risks: int
    critical_risks: int
    anomaly_events_count: int
    predicted_failures: int
    average_confidence: float
    predictions_by_service: dict[str, int]
    predictions_by_metric: dict[str, int]
    predictions_by_risk: dict[str, int]


class ServiceRiskItem(BaseModel):
    service: str
    region: str
    risk_level: str
    failure_probability: float
    active_predictions_count: int


class InfrastructureRiskHeatmapResponse(BaseModel):
    items: list[ServiceRiskItem]


class PredictionAnalyzeRequest(BaseModel):
    services: list[str] | None = Field(
        default=None, description="Optional list of specific services to analyze"
    )
    lookback_hours: int = Field(
        default=24, ge=1, le=168, description="Telemetry lookback window in hours"
    )
    telemetry_map: dict[str, list[float]] | None = Field(
        default=None, description="Optional explicit telemetry series to evaluate"
    )


class CreateIncidentFromPredictionRequest(BaseModel):
    severity: str | None = Field(default=None, description="Optional override severity")
    title: str | None = Field(default=None, description="Optional incident title")
